### PR TITLE
WT-11081 Collect and ingest code complexity Evergreen metrics

### PR DESCRIPTION
**Pull request by raviprakashgiri29**

### CODE REVIEW

File: test/evergreen.yml [hunk 1]
```diff
@@ -4568,6 +4568,14 @@ tasks:
                 cat $t && rm $t
             fi
 
+            # Code complexity statistics
+
```
Comment (lukech) [inline test/evergreen.yml:4572]: The extra blank line can be saved. Technically the above lines are also "Code complexity statistics" related. Suggest making the comment more specific/descriptive.

File: test/evergreen.yml [hunk 2]
```diff
@@ -4568,6 +4568,14 @@ tasks:
...
+            python3 "../metrixplusplus/metrix++.py" view --format=python > code_complexity_summary.txt
+            python3 "../metrixplusplus/metrix++.py" export --db-file=metrixpp.db > metrixpp.csv
```
Comment (lukech) [inline test/evergreen.yml:4574]: python3 --> python to be consistent with existing metrix++.py calls above.

Comment (lukech) [reply test/evergreen.yml:4574]: code_complexity_summary.txt --> code_complexity_summary.json to reflect the format of the output.

File: test/evergreen.yml [hunk 3]
```diff
@@ -4568,6 +4568,14 @@ tasks:
...
+
+            cd ../test/evergreen
```
Comment (lukech) [inline test/evergreen.yml:4576]: Is it necessary to change directory here?

File: test/evergreen/code_complexity_analysis.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
+# importing the module
+import argparse
+import json
+import csv
+import os.path
+import ast
+from heapq import nlargest
+import pandas as pd
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:8]: Need a pip install in a virtual environment.

File: test/evergreen/code_complexity_analysis.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
+# importing the module
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:1]: This comment can be saved. Instead, the shebang line can be added.

File: test/evergreen/code_complexity_analysis.py (hunk 4 above)
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:8]: Suggest sorting the import lines in alphabetical order.

File: test/evergreen/code_complexity_analysis.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+from collections import defaultdict
+    
+class CodeComplexity:
+    dataFrame = pd.DataFrame()
+    topNRegions = 5
+    def __init__(self):
+        self.parser = argparse.ArgumentParser()
+        self.parser.add_argument('-s', '--summary', required=True, help='Path to the complexity summary data file')
+        self.parser.add_argument('-o', '--outfile', help='Path of the file to write test output to')
+        self.parser.add_argument('-d', '--data_file', help='Code complexity data file')
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:18]: It's a bit unusual to have the arg handling in the __init__ method. Should they be moved to a main() function?

File: test/evergreen/code_complexity_analysis.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import csv
+import os.path
+import ast
+from heapq import nlargest
+import pandas as pd 
+from collections import defaultdict
+    
+class CodeComplexity:
+    dataFrame = pd.DataFrame()
+    topNRegions = 5
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:13]: Should they go to the __init__ method?

File: test/evergreen/code_complexity_analysis.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import argparse
+import json
+import csv
+import os.path
+import ast
+from heapq import nlargest
+import pandas as pd 
+from collections import defaultdict
+    
+class CodeComplexity:
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:11]: I don't see much value in having a class defined except for grouping the related functions together, which the script itself has achieved.

File: test/evergreen/code_complexity_analysis.py (hunk 6 above)
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:18]: If only csv files are expected, suggest mentioning it in the help message.

File: test/evergreen/code_complexity_analysis.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    dataFrame = pd.DataFrame()
+    topNRegions = 5
+    def __init__(self):
+        self.parser = argparse.ArgumentParser()
+        self.parser.add_argument('-s', '--summary', required=True, help='Path to the complexity summary data file')
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:16]: If deciding to change txt to json for the summary file, suggest mentioning it in the help message (to distinguish it from the other data file "-d")

File: test/evergreen/code_complexity_analysis.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+        self.parser.add_argument('-o', '--outfile', help='Path of the file to write test output to')
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:17]: Remove "test" or replace it with "analysis".

File: test/evergreen/code_complexity_analysis.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    def run(self):
+        self.args = self.parser.parse_args()
+        self.dataFrame = pd.read_csv(self.args.data_file)
+
+        self.get_atlas_compatible_code_statistics(self.args.summary, self.args.data_file, self.args.outfile)
+
+    # Generate the Atlas compatible format report.
+    def get_atlas_compatible_code_statistics(self, summaryFile, dataFile, outfile='atlas_code_statistics.json'):
+        atlas_format = {
+                    'Test Name': "Code Statistics",
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:29]: Should the test name be "Code Complexity" to be distinguished from the existing "Code Coverage" test name?

File: test/evergreen.yml [hunk 12]
```diff
@@ -4568,6 +4568,14 @@ tasks:
...
+            python3 code_complexity_analysis.py -s ../../src/code_complexity_summary.txt -d ../../src/metrixpp.csv -o atlas_code_statistics.json
```
Comment (lukech) [inline test/evergreen.yml:4577]: atlas_code_statistics.json --> atlas_code_complexity.json

File: test/evergreen/code_complexity_analysis.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+                }
+
+        dir_name = os.path.dirname(outfile)
+        if dir_name:
+            os.makedirs(dir_name, exist_ok=True)
+
+        with open(outfile, 'w') as outfile:
+            json.dump(atlas_format, outfile, indent=4)
+
+    def get_code_statistics(self, summaryFile, dataFile):
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:41]: get_code_statistics --> get_code_complexity

File: test/evergreen/code_complexity_analysis.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        # Select column 'std.code.complexity:cyclomatic' from the dataframe
+        column = data[column_name]
+        # Get count of values greater than max_limit in the column 'std.code.complexity:cyclomatic' 
+        count = column[column > max_limit].count()
+        #Pandas aggregation functions (like sum, count and mean) returns a NumPy int64 type number
+        #not a Python integer.
+        # Object of type int64 is not JSON serializable | pandas aggregation functions and json.dumps() error
+        return (int(count))
+
+    def get_average(self, complexity_summary_path):
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:106]: complexity_summary_path --> complexity_summary_file

File: test/evergreen/code_complexity_analysis.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        with open(complexity_summary_path) as f:
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:107]: Specify "-r" as the 2nd arg for the open call.

File: test/evergreen/code_complexity_analysis.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+            data = f.read()
+
+        d = ast.literal_eval(data)
+        new_dictionary = d["view"]
+        return (new_dictionary[0]['data']['aggregated-data']['std.code.complexity']['cyclomatic']['avg'])
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:112]: It's hard to follow without a sample structure of the data. Do you having one that can be added as a comment?

File: test/evergreen/code_complexity_analysis.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+            atlas_format['name'] = row["region"]
+            atlas_format['value']= row["std.code.complexity:cyclomatic"]
+            resultList.append(atlas_format)
+
+        print ("Region List ")
+        print (resultList)
+        return (resultList)
+
+    def get_complexity_ranges_list(self, dataFile):
+        ranges_list = [20, 50, 90]
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:79]: It feels worthwhile to have a docstring describing we are aimed to get 3 x ranges of complexity data, i.e. above 20, above 50, and above 90.

File: test/evergreen/code_complexity_analysis.py [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        complexity_regions_dict['value'] = self.get_region_list(dataFile)
+
+        resultList.append(complexity_avg_dict)
+        resultList.append(complexity_ranges_dict)
+        resultList.append(complexity_regions_dict)
+
+        return (resultList)
+
+    def get_region_list(self, dataFile):
+        #data = pd.read_csv(dataFile)
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:62]: Can be removed?

File: test/evergreen/code_complexity_analysis.py [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+        complexity_regions_dict['name'] = 'Top ' + str(self.topNRegions) + ' Regions'
+        complexity_regions_dict['value'] = self.get_region_list(dataFile)
+
+        resultList.append(complexity_avg_dict)
+        resultList.append(complexity_ranges_dict)
+        resultList.append(complexity_regions_dict)
+
+        return (resultList)
+
+    def get_region_list(self, dataFile):
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:61]: Worth a docstring describing we are aimed to collect the top 5 regions/functions with the highest cyclomatic complexity values.

File: test/evergreen/code_complexity_analysis.py [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+        #data = pd.read_csv(dataFile)
+        list = []
+        # extracting greatest 5
+        large5 = self.dataFrame.nlargest(self.topNRegions, 'std.code.complexity:cyclomatic')
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:65]: large5 --> top5

**Review by lukech — CHANGES_REQUESTED** [review]
In the design doc, there are mentions for using the “--max-limit” command option to retrieve the regions of above 20/50/90 ranges complexity data, and using the “--hotspot” command option to retrieve the top 5 regions with the highest cyclomatic complexity values. Did you try that path?

File: test/evergreen.yml (hunk 1 above)
Comment (raviprakashgiri29) [reply test/evergreen.yml]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen.yml (hunk 2 above)
Comment (raviprakashgiri29) [reply test/evergreen.yml]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen.yml (hunk 3 above)
Comment (raviprakashgiri29) [reply test/evergreen.yml]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5 to use the relative path

File: test/evergreen/code_complexity_analysis.py (hunk 4 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 5 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 4 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 6 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 7 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 6 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 9 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 10 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 11 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen.yml (hunk 12 above)
Comment (raviprakashgiri29) [reply test/evergreen.yml]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 13 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 14 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 15 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 16 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 17 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 18 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 19 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

File: test/evergreen/code_complexity_analysis.py (hunk 20 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 7cc186a02197e388c9a2fcaa97331d24c84376a5

**Comment by raviprakashgiri29** [comment]
> In the design doc, there are mentions for using the “--max-limit” command option to retrieve the regions of above 20/50/90 ranges complexity data, and using the “--hotspot” command option to retrieve the top 5 regions with the highest cyclomatic complexity values. Did you try that path?

Yes, I tried using the command options to metrix++ but outputs the data in the CSV format which I have to parse it again so I parse the super set summary file into a data frame and extract the infirmation I need.

File: test/evergreen/code_complexity_analysis.py [hunk 21]
```diff
@@ -0,0 +1,126 @@
+import argparse
+import ast
+from collections import defaultdict
+import csv
+from heapq import nlargest
+import json
+import os.path
+import pandas as pd
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:8]: Can the "from .. import .." lines be grouped and put after those "import .." lines, with a blank line separation in between?

File: test/evergreen/code_complexity_analysis.py [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+    ''' Retrieve topNRegions/functions with the highest cyclomatic complexity values. '''
+    top5 = dataFrame.nlargest(topNRegions, 'std.code.complexity:cyclomatic')
+    atlasFormat = {}
+    for index, row in top5.iterrows():
+        atlasFormat[row["region"]] = row["std.code.complexity:cyclomatic"]
+
+    return (atlasFormat)
+
+def get_complexity_ranges_list(dataFile):
+    ''' Retrieve 3 set of cyclomatic complexity ranges, above 30, above 50 and above 90. '''
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:46]: above 30 --> above 20

File: test/evergreen/code_complexity_analysis.py [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+    complexityDict = {}
+    complexityDict['Average'] = get_average(summaryFile)
+    complexityDict['Stats Ranges'] = get_complexity_ranges_list(dataFile)
+    complexityDict['Top ' + str(topNRegions) + ' Regions'] = get_region_list(dataFile)
+    
+    resultList.append(complexityDict)
+    return (resultList)
+
+def get_region_list(dataFile):
+    ''' Retrieve topNRegions/functions with the highest cyclomatic complexity values. '''
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:37]: Is a space missed after `topN`? 
topN --> top5 if there's no plan to make N as a function arg.

File: test/evergreen.yml [hunk 24]
```diff
@@ -4568,6 +4570,22 @@ tasks:
                 cat $t && rm $t
             fi
 
+            # Generate the code complexity statistics and upload the analysis into Atlas charts
+            virtualenv -p python3 venv
+            source venv/bin/activate
+            pip3 install pandas
```
Comment (lukech) [inline test/evergreen.yml:4576]: Suggest moving the 3 lines down (in front of line 4581) as that's where the pandas module is needed.

File: test/evergreen.yml [hunk 25]
```diff
@@ -4568,6 +4570,22 @@ tasks:
...
+
+            python3 "../metrixplusplus/metrix++.py" view --format=python > ../code_statistics_report/code_complexity_summary.json
+            python3 "../metrixplusplus/metrix++.py" export --db-file=metrixpp.db > ../code_statistics_report/metrixpp.csv
+
+            python3 ../test/evergreen/code_complexity_analysis.py -s ../code_statistics_report/code_complexity_summary.json -d ../code_statistics_report/metrixpp.csv -o ../code_statistics_report/atlas_out_code_complexity.json
+      - func: "upload stats to atlas"
+        vars:
+          stats_dir: ./wiredtiger/code_statistics_report
+          test-name: code_complexity
+          # Override the collection name to use the code complexity collection.
```
Comment (lukech) [inline test/evergreen.yml:4586]: I think we should remove the default setting for collection value at the code statistics builder level, and let each task take care of its own collection value setting. And thus remove the need for this comment line.

File: test/evergreen/code_complexity_analysis.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+    top5 = dataFrame.nlargest(topNRegions, 'std.code.complexity:cyclomatic')
+    atlasFormat = {}
+    for index, row in top5.iterrows():
+        atlasFormat[row["region"]] = row["std.code.complexity:cyclomatic"]
+
+    return (atlasFormat)
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:43]: Just curious if the parentheses are necessary here.

File: test/evergreen/code_complexity_analysis.py [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+import ast
+from collections import defaultdict
+import csv
+from heapq import nlargest
+import json
+import os.path
+import pandas as pd
+
+def get_atlas_compatible_code_statistics(summaryFile, dataFile, outfile='atlas_code_complexity.json'):
+    ''' Generate the Atlas compatible format report. '''
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:11]: Suggest adding a blank line after the docstring (for ease of reading). This is a general comment applied to the below functions too.

File: test/evergreen/code_complexity_analysis.py [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+    ''' Retrieve 3 set of cyclomatic complexity ranges, above 30, above 50 and above 90. '''
+    rangesList = [20, 50, 90]
+    atlasFormat = {}
+    for i in rangesList:
+        aboveRangeString = 'Above ' + str(i)
+        atlasFormat[aboveRangeString] = get_complexity_max_limit(dataFile, i)
+
+    return (atlasFormat)
+
+def get_complexity_max_limit(complexity_data_path: str, max_limit:int):
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:55]: complexity_data_path --> complexity_data_file

Comment (lukech) [reply test/evergreen/code_complexity_analysis.py:55]: Adding a space after `max_limit:` for consistency.

File: test/evergreen/code_complexity_analysis.py [hunk 29]
```diff
@@ -0,0 +1,126 @@
...
+    ''' Retrieve the number of cyclomatic complexity values above the max_limit. '''
+    columnName = 'std.code.complexity:cyclomatic'
+    # Select column 'std.code.complexity:cyclomatic' from the dataframe
+    column = dataFrame[columnName]
+    # Get count of values greater than max_limit in the column 'std.code.complexity:cyclomatic'
+    count = column[column > max_limit].count()
+    """
+    Pandas aggregation functions (like sum, count and mean) returns a NumPy int64 type number not a Python integer.
+    Object of type int64 is not JSON serializable so need to convert into an int.
+    """
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:65]: This looks like a docstring. Suggest using `#` at the beginning of the 2 comment lines instead.

File: test/evergreen/code_complexity_analysis.py [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+    return (int(count))
+
+def get_average(complexity_summary_file):
+    ''' Retrieve the cyclomatic code complexity value '''
+    with open(complexity_summary_file, 'r') as f:
+        data = f.read()
+
+    # Below is the format of complexity_summary_file, we are intersted in extracting the "avg" value of std.code.complexity.
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:73]: Type "intersted"

File: test/evergreen/code_complexity_analysis.py [hunk 31]
```diff
@@ -0,0 +1,126 @@
...
+                        },
+                    },
+                    "file-data": {},
+                    "subdirs": [ ],
+                    "subfiles": [ ],
+                }
+            }
+        ]
+    }
+    """
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:104]: Could this section be moved up into the docstring?

File: test/evergreen/code_complexity_analysis.py [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+
+    d = ast.literal_eval(data)
+    newDictionary = d["view"]
+
+    return (newDictionary[0]['data']['aggregated-data']['std.code.complexity']['cyclomatic']['avg'])
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:109]: It's safer to include this section of code in a try exception block, to capture miss-formated data files. 
Also, suggest returning a variable instead of the super-long statement.

File: test/evergreen/code_complexity_analysis.py [hunk 33]
```diff
@@ -0,0 +1,126 @@
...
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-s', '--summary', required=True, help='Path of the complexity summary file in the json format')
+    parser.add_argument('-o', '--outfile', help='Path of the file to write analysis output to')
+    parser.add_argument('-d', '--data_file', help='Code complexity data file in the csv format')
+
+    args = parser.parse_args()
+    global dataFrame
+    global topNRegions
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:119]: Is it necessary to use global variables? 
Looks like the dataFrame variable is red in 2 x functions, can we move lines 121-122 closer to where it's directly used e.g. within `get_code_complexity()` and pass it as an arg for the reader functions?

File: test/evergreen/code_complexity_analysis.py (hunk 22 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen.yml [hunk 34]
```diff
@@ -4568,6 +4570,22 @@ tasks:
                 cat $t && rm $t
             fi
 
+            # Generate the code complexity statistics and upload the analysis into Atlas charts
```
[the remark this answers is not in the document]
Comment (raviprakashgiri29) [reply test/evergreen.yml]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen.yml (hunk 25 above)
Comment (raviprakashgiri29) [reply test/evergreen.yml:4586]: Yes, I thought about it but the collection `CodeCoverage` is being used by 50+ functions so I implemented this way to use the `CodeCoverage` collection as a default setting and override it once in the `cyclomatic-complexity`

File: test/evergreen/code_complexity_analysis.py (hunk 26 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py:52]: It is just a coding style I used in this script.

File: test/evergreen/code_complexity_analysis.py (hunk 27 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py:13]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen/code_complexity_analysis.py (hunk 28 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen/code_complexity_analysis.py (hunk 29 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen/code_complexity_analysis.py (hunk 30 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen/code_complexity_analysis.py (hunk 31 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 3154760ada6586d59d072c5df2e956a40df9eba9

File: test/evergreen/code_complexity_analysis.py (hunk 33 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: I moved the initializations of the `dataFrame ` and `topNRegions` to the `get_code_complexity` method but still kept it global instead to pass it an argument.

File: test/evergreen/code_complexity_analysis.py [hunk 35]
```diff
@@ -0,0 +1,137 @@
...
+    count = column[column > max_limit].count()
+
+    # Pandas aggregation functions (like sum, count and mean) returns a NumPy int64 type number not a Python integer.
+    # Object of type int64 is not JSON serializable so need to convert into an int.
+
+    return (int(count))
+
+def get_average(complexity_summary_file):
+    ''' Retrieve the cyclomatic code complexity value '''
+    """
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:81]: The docstring markers """/''' are duplicated. One pair should be good enough.

File: test/evergreen/code_complexity_analysis.py [hunk 36]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        ]
+    }
+    """
+
+    with open(complexity_summary_file, 'r') as f:
+        data = f.read()
+
+    d = ast.literal_eval(data)
+    newDictionary = d["view"]
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:118]: I think this line should be put into the try/except block as it's susceptible to hit an exception if the data is not formatted as expected.

File: test/evergreen/code_complexity_analysis.py [hunk 37]
```diff
@@ -0,0 +1,137 @@
...
+
+    try:
+        average = newDictionary[0]['data']['aggregated-data']['std.code.complexity']['cyclomatic']['avg']
+    except Exception as e:
+        sys.exit(f"The summary file '{complexity_summary_file}' does not have cyclomatic code complexity average value. {e.text}")
```
Comment (lukech) [inline test/evergreen/code_complexity_analysis.py:123]: While the case described by the text is one scenario an exception could hit, there are other cases. Suggest making the exit text a bit more general e.g. "Could not retrieve the cyclomatic code complexity average value. Please make sure the summary file is properly formatted".

File: test/evergreen/code_complexity_analysis.py (hunk 33 above)
Comment (lukech) [reply test/evergreen/code_complexity_analysis.py]: There are some caveats about using global variables in Python. I'd rather avoid using them unless necessary. Cann't you pass the 2 variables as function args to the functions that make use of them (e.g. `get_code_complexity`) and make them non-global?

Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 19d2a584b833bc89692c079e17033ea0c4609175. Removed using global variables and passing it as an function argument.

File: test/evergreen/code_complexity_analysis.py (hunk 35 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py:78]: Fixed in 19d2a584b833bc89692c079e17033ea0c4609175

File: test/evergreen/code_complexity_analysis.py (hunk 36 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 19d2a584b833bc89692c079e17033ea0c4609175

File: test/evergreen/code_complexity_analysis.py (hunk 37 above)
Comment (raviprakashgiri29) [reply test/evergreen/code_complexity_analysis.py]: Fixed in 19d2a584b833bc89692c079e17033ea0c4609175

**Review by lukech — APPROVED** [review]
LGTM, thanks.
